### DISCUSSION TITLE
How to use GenerateRedirects module?

### DISCUSSION THREAD

**Discussion by MetaFight** [statiqdev/Discussions#219] [Q&A] [answered]
Hi everyone,

I'm using Statiq.Web but I opted to use the `.CreateDefault` bootsrap extension rather than the `.CreateWeb` one.
I've managed to set up a pipeline where my documents and assets got processed and rendered the way I want, but whenever I try to use the `GenerateRedirects` module it will generate the redirect but also *delete the original document*.

If I switch to boostrapping with `.CreateWeb` then my documents are generated, but my redirects aren't.

I've tried using `GenerateRedirects` as a Process module and as a Post Process module and neither work.
Looking at the module's source I see that there is no logic to preserve the source document.

What am I doing wrong?  How should I be using this module so that the redirects are generated while also preserving the original document?

**Comment by daveaglick** [comment]
Just have my phone without a laptop the next couple days, but I might be able to assist before I get back in town (and can definitely help more once I've got my laptop back). This is all from memory, so take it with a grain of salt, but hopefully I can help unblock you.

I think the problem might be that the `GenerateRedirects` module "eats" the input documents. It's only job is to output document(s) that represent the redirect files, whether those are meta-refresh placeholders or something else like a Netlify redirect file.

To ensure the original documents that the `GenerateRedirects` module reads are also output, you'll want to run the redirect module not "inline" with your main document processing. One way to do that is stick it in a separate pipeline that reads the documents from your "main" pipeline and is just responsible for creating and outputting the redirect file(s). There are also modules designed to set up "branches" that might work as well to execute the redirects module and then combine it's outputs with the main outputs (see the modules in https://github.com/statiqdev/Statiq.Framework/tree/main/src/core/Statiq.Core/Modules/Control).

Does that make sense and seem like I'm on the right track? Let me know how you get along.

**Comment by MetaFight** [comment, marked as answer]
So, I had considered "splitting" my pipeline into parallel branches like you suggest, but I couldn't figure out how.

Your link to the Control modules did help me find a solution in the end.  thanks!

What worked for me is to wrap the `GenerateRedirects` module in a `ConcatDocuments` module as such:
```cs
this.ProcessModules = new()
{
   // ...
    new ConcatDocuments(new GenerateRedirects()),
   // ...
}
```
